src/hotspot/share/ci/ciArray.cpp

```diff
@@ -94,9 +94,11 @@ ciConstant ciArray::element_value_impl(BasicType elembt,
 // Returns T_ILLEGAL if there is no element at the given index.
 ciConstant ciArray::element_value(int index) {
   BasicType elembt = element_basic_type();
+  ciConstant value;
   GUARDED_VM_ENTRY(
-    return element_value_impl(elembt, get_arrayOop(), index);
+    value = element_value_impl(elembt, get_arrayOop(), index);
   )
+  return check_constant_value_cache(index, value);
```

Comment:
It still unconditionally goes through VM transition on every call. If you check the cache first and call `element_value_impl()` only when there's a miss, then you can save on state transition when the cache is populated. But you would need additional call to update the cache.

Reply:
Or you could move both checks into `element_value_impl()` along with `GUARDED_VM_ENTRY` (and convert it into `VM_ENTRY_MARK`).